gcloud-java-core/src/main/java/com/google/gcloud/BaseServiceException.java

```diff
@@ -16,26 +16,80 @@
 
 package com.google.gcloud;
 
+import com.google.api.client.googleapis.json.GoogleJsonError;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+
+import java.io.IOException;
+import java.net.SocketTimeoutException;
+import java.util.Collections;
+import java.util.Set;
+
 /**
  * Base class for all service exceptions.
  */
 public class BaseServiceException extends RuntimeException {
 
   private static final long serialVersionUID = 5028833760039966178L;
+  public static final int UNKNOWN_CODE = 0;
 
   private final int code;
   private final boolean retryable;
+  private final boolean idempotent;
+
+  public BaseServiceException(IOException exception, boolean idempotent) {
+    super(exception.getMessage(), exception);
+    this.code = statusCode(exception);
+    this.idempotent = idempotent;
+    this.retryable = idempotent && isRetryable(exception);
+  }
+
+  public BaseServiceException(GoogleJsonError error, boolean idempotent) {
+    super(error.getMessage());
+    this.code = error.getCode();
+    this.idempotent = idempotent;
+    this.retryable = idempotent && isRetryable(error);
+  }
 
-  public BaseServiceException(int code, String message, boolean retryable) {
+  public BaseServiceException(int code, String message, boolean retryable, boolean idempotent) {
     super(message);
     this.code = code;
-    this.retryable = retryable;
+    this.idempotent = idempotent;
+    this.retryable = idempotent && retryable;
   }
 
-  public BaseServiceException(int code, String message, boolean retryable, Exception cause) {
+  public BaseServiceException(int code, String message, boolean retryable, boolean idempotent,
+      Exception cause) {
     super(message, cause);
     this.code = code;
-    this.retryable = retryable;
+    this.idempotent = idempotent;
+    this.retryable = idempotent && retryable;
+  }
+
+  protected Set<Integer> retryableCodes() {
+    return Collections.emptySet();
+  }
+
+  protected int statusCode(IOException exception) {
+    if (exception instanceof GoogleJsonResponseException) {
```

Comment:
> 1) 2) and 3) What holds me from including (and relying on) 

Though consistency for reason naming across services would be nice it should not be a problem as we treat them just as string and it is the service that defines which strings indicates a retryable exception.

> 1) Isn't the reason stored in `GoogleJsonError.getErrors().get(0).getReason()`?

Yes, that is correct we should use that one instead of `GoogleJsonError.getMessage()`

> 4) This is does not seem to be what we do in Storage

I agree, consistency would be prefered and we can actually take the extra step
and throw the DatastoreException from the RPC layer (initially for Datastore there was a complete separation of the API and SPI including using API from SPI [which in most parts should be OK).